diff --git a/Source/WPEFramework/PluginHost.cpp b/Source/WPEFramework/PluginHost.cpp
index e14d162..c514794 100644
--- a/Source/WPEFramework/PluginHost.cpp
+++ b/Source/WPEFramework/PluginHost.cpp
@@ -23,6 +23,7 @@
 #include <dlfcn.h> // for dladdr
 #include <syslog.h>
 #endif
+#include <signal.h>
 
 MODULE_NAME_DECLARATION(BUILD_REFERENCE)
 
@@ -264,7 +265,7 @@ namespace PluginHost {
         //Set our Logging Mask and open the Log
         setlogmask(LOG_UPTO(LOG_NOTICE));
         openlog(argv[0], LOG_PID, LOG_USER);
-
+        signal(SIGCHLD,SIG_IGN);
         setsid();
 #endif
 
