From 7aa1b22656cbce455d80a42e04f3692174e92dc2 Mon Sep 17 00:00:00 2001
From: Arun P Madhavan <arun_madhavan@comcast.com>
Date: Thu, 17 Sep 2020 07:38:52 +0000
Subject: [PATCH] lower version in FindWPE for libwpe-0.2

---
 cmake/FindWPE.cmake | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/cmake/FindWPE.cmake b/cmake/FindWPE.cmake
index a062481..7f4ffd2 100644
--- a/cmake/FindWPE.cmake
+++ b/cmake/FindWPE.cmake
@@ -29,7 +29,7 @@
 # ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 find_package(PkgConfig)
-pkg_check_modules(PC_WPE QUIET wpe-1.0)
+pkg_check_modules(PC_WPE QUIET wpe-0.2)
 
 find_path(WPE_INCLUDE_DIRS
     NAMES wpe/wpe.h
@@ -37,7 +37,7 @@ find_path(WPE_INCLUDE_DIRS
 )
 
 find_library(WPE_LIBRARIES
-    NAMES wpe-1.0
+    NAMES wpe-0.2
     HINTS ${PC_WPE_LIBDIR} ${PC_WPE_LIBRARY_DIRS}
 )
 
-- 
2.24.0

