diff --git a/Clients/Makefile b/Clients/Makefile
index 383af31..52a2462 100755
--- a/Clients/Makefile
+++ b/Clients/Makefile
@@ -46,10 +46,10 @@ build:
 	mkdir build
 
 build/dns-sd: build dns-sd.c ClientCommon.c
-	$(CC) $(filter %.c %.o, $+) $(LIBS) -I../mDNSShared -Wall -o $@
+	$(CC) $(CFLAGS_EXTRA) $(filter %.c %.o, $+) $(LIBS) -I../mDNSShared -Wall -o $@
 
 build/dns-sd64: build dns-sd.c ClientCommon.c
-	$(CC) $(filter %.c %.o, $+) $(LIBS) -I../mDNSShared -Wall -o $@ -m64
+	$(CC) $(CFLAGS_EXTRA) $(filter %.c %.o, $+) $(LIBS) -I../mDNSShared -Wall -o $@ -m64
 
 # Note, we can make a 'fat' version of dns-sd using 'lipo', as shown below, but we
 # don't, because we don't want or need a 'fat' version of dns-sd, because it will
